code fix for copper cable modules with flat memory

qinchuanares opened the pull request:
This PR aims to fix issues on copper cable modules with flat memory

#### Description
Copper cable modules only have flat memory. Some APIs in soic_xcvr\api\public\cmis.py do not handle flat memory gracefully and raise errors. This PR aims to fix those issues and deal with flat memory without errors.

#### Motivation and Context
See descrition

#### How Has This Been Tested?
test_cmis.py cover the flat memory case

#### Additional Information (Optional)

Example CLI output:

root@sonic:~# sfputil show eeprom -d -p Ethernet16
WARNING: api_use_sfpoptoe attr type changed: old <class 'bool'>, new <class 'str'>
Ethernet16: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: MPO 1x12
        Encoding: N/A
        Extended Identifier: Power Class 6 (12.0W Max)
        Extended RateSelect Compliance: N/A
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Length Cable Assembly(m): 0.0
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: Single Mode Fiber (SMF)
        Vendor Date Code(YYYY-MM-DD Lot): 2021-05-24
        Vendor Name: MOLEX
        Vendor OUI: 00-09-3a
        Vendor PN: 1064281000
        Vendor Rev: 1.1
        Vendor SN: G21211B3113
        ChannelMonitorValues:
                RX1Power: 0.0001dBm
                RX2Power: 0.0001dBm
                RX3Power: 0.0001dBm
                RX4Power: 0.0001dBm
                RX5Power: 0.0dBm
                RX6Power: 0.0dBm
                RX7Power: 0.0dBm
                RX8Power: 0.0dBm
                TX1Bias: 0.0mA
                TX1Power: 0.0001dBm
                TX2Bias: 0.0mA
                TX2Power: 0.0001dBm
                TX3Bias: 0.0mA
                TX3Power: 0.0001dBm
                TX4Bias: 0.0mA
                TX4Power: 0.0001dBm
                TX5Bias: 0.0mA
                TX5Power: 0.0dBm
                TX6Bias: 0.0mA
                TX6Power: 0.0dBm
                TX7Bias: 0.0mA
                TX7Power: 0.0dBm
                TX8Bias: 0.0mA
                TX8Power: 0.0dBm
        ChannelThresholdValues:
                RxPowerHighAlarm  : 5.0dBm
                RxPowerHighWarning: 4.5dBm
                RxPowerLowAlarm   : -8.901dBm
                RxPowerLowWarning : -6.899dBm
                TxBiasHighAlarm   : 95.0mA
                TxBiasHighWarning : 88.0mA
                TxBiasLowAlarm    : 8.0mA
                TxBiasLowWarning  : 10.0mA
                TxPowerHighAlarm  : 5.0dBm
                TxPowerHighWarning: 4.5dBm
                TxPowerLowAlarm   : -3.401dBm
                TxPowerLowWarning : -2.901dBm
        ModuleMonitorValues:
                Temperature: 21.004C
                Vcc: 3.377Volts
        ModuleThresholdValues:
                TempHighAlarm  : 75.0C
                TempHighWarning: 72.0C
                TempLowAlarm   : -5.0C
                TempLowWarning : -2.0C
                VccHighAlarm   : 3.51Volts
                VccHighWarning : 3.495Volts
                VccLowAlarm    : 3.1Volts
                VccLowWarning  : 3.115Volts

Example API output:

get_transceiver_info

{'type': 'QSFP-DD Double Density 8X Pluggable Transceiver'
 'type_abbrv_name': 'QSFP-DD'
 'hardware_rev': '1.1'
 'serial': 'G21211B3113     '
 'manufacturer': 'MOLEX           '
 'model': '1064281000      '
 'connector': 'MPO 1x12'
 'encoding': 'N/A'
 'ext_identifier': 'Power Class 6 (12.0W Max)'
 'ext_rateselect_compliance': 'N/A'
 'cable_type': 'Length Cable Assembly(m)'
 'cable_length': 0.0
 'nominal_bit_rate': 0
 'specification_compliance': 'Single Mode Fiber (SMF)'
 'vendor_date': '2021-05-24   '
 'vendor_oui': '00-09-3a'
 'application_advertisement': 'N/A'
 'host_electrical_interface': '400GAUI-8 C2M (Annex 120E)'
 'media_interface_code': '400GBASE-DR4 (Cl 124)'
 'host_lane_count': 8
 'media_lane_count': 4
 'host_lane_assignment_option': 1
 'media_lane_assignment_option': 1
 'active_apsel_hostlane1': 2
 'active_apsel_hostlane2': 2
 'active_apsel_hostlane3': 2
 'active_apsel_hostlane4': 2
 'active_apsel_hostlane5': 2
 'active_apsel_hostlane6': 2
 'active_apsel_hostlane7': 2
 'active_apsel_hostlane8': 2
 'media_interface_technology': '1310 nm DFB'
 'vendor_rev': '11'
 'cmis_rev': '4.0'
 'active_firmware': '3.1'
 'inactive_firmware': '3.1'}



get_transceiver_bulk_status

{'rx_los': False
 'tx_fault': 'N/A'
 'tx_disabled_channel': 0
 'temperature': 20.926
 'voltage': 3.375
 'tx1disable': False
 'tx1bias': 0.0
 'rx1power': -40.0
 'tx1power': -40.0
 'tx2disable': False
 'tx2bias': 0.0
 'rx2power': -40.0
 'tx2power': -40.0
 'tx3disable': False
 'tx3bias': 0.0
 'rx3power': -40.0
 'tx3power': -40.0
 'tx4disable': False
 'tx4bias': 0.0
 'rx4power': -40.0
 'tx4power': -40.0
 'tx5disable': False
 'tx5bias': 0.0
 'rx5power': -inf
 'tx5power': -inf
 'tx6disable': False
 'tx6bias': 0.0
 'rx6power': -inf
 'tx6power': -inf
 'tx7disable': False
 'tx7bias': 0.0
 'rx7power': -inf
 'tx7power': -inf
 'tx8disable': False
 'tx8bias': 0.0
 'rx8power': -inf
 'tx8power': -inf
 'laser_temperature': 0.0
 'prefec_ber': 0.0}



get_transceiver_threshold_info

{'temphighalarm': 75.0
 'templowalarm': -5.0
 'temphighwarning': 72.0
 'templowwarning': -2.0
 'vcchighalarm': 3.51
 'vcclowalarm': 3.1
 'vcchighwarning': 3.495
 'vcclowwarning': 3.115
 'rxpowerhighalarm': 5.0
 'rxpowerlowalarm': -8.901
 'rxpowerhighwarning': 4.5
 'rxpowerlowwarning': -6.899
 'txpowerhighalarm': 5.0
 'txpowerlowalarm': -3.401
 'txpowerhighwarning': 4.5
 'txpowerlowwarning': -2.901
 'txbiashighalarm': 380.0
 'txbiaslowalarm': 32.0
 'txbiashighwarning': 352.0
 'txbiaslowwarning': 40.0
 'lasertemphighalarm': 0.0
 'lasertemplowalarm': 0.0
 'lasertemphighwarning': 0.0
 'lasertemplowwarning': 0.0
 'prefecberhighalarm': 7.699999999999999e-20
 'prefecberlowalarm': 1.868e-12
 'prefecberhighwarning': 1.368e-13
 'prefecberlowwarning': 3.2e-19}



get_transceiver_status

{'module_state': 'ModuleLowPwr'
 'module_fault_cause': 'No Fault detected'
 'datapath_firmware_fault': False
 'module_firmware_fault': False
 'module_state_changed': False
 'temphighalarm_flag': False
 'templowalarm_flag': False
 'temphighwarning_flag': False
 'templowwarning_flag': False
 'vcchighalarm_flag': False
 'vcclowalarm_flag': False
 'vcchighwarning_flag': False
 'vcclowwarning_flag': False
 'lasertemphighalarm_flag': False
 'lasertemplowalarm_flag': False
 'lasertemphighwarning_flag': False
 'lasertemplowwarning_flag': False
 'DP1State': 'DataPathDeactivated'
 'DP2State': 'DataPathDeactivated'
 'DP3State': 'DataPathDeactivated'
 'DP4State': 'DataPathDeactivated'
 'DP5State': 'DataPathDeactivated'
 'DP6State': 'DataPathDeactivated'
 'DP7State': 'DataPathDeactivated'
 'DP8State': 'DataPathDeactivated'
 'txoutput_status1': False
 'txoutput_status2': False
 'txoutput_status3': False
 'txoutput_status4': False
 'txoutput_status5': False
 'txoutput_status6': False
 'txoutput_status7': False
 'txoutput_status8': False
 'rxoutput_status_hostlane1': False
 'rxoutput_status_hostlane2': False
 'rxoutput_status_hostlane3': False
 'rxoutput_status_hostlane4': False
 'rxoutput_status_hostlane5': False
 'rxoutput_status_hostlane6': False
 'rxoutput_status_hostlane7': False
 'rxoutput_status_hostlane8': False
 'txfault1': 'N/A'
 'txfault2': 'N/A'
 'txfault3': 'N/A'
 'txfault4': 'N/A'
 'txfault5': 'N/A'
 'txfault6': 'N/A'
 'txfault7': 'N/A'
 'txfault8': 'N/A'
 'txlos_hostlane1': False
 'txlos_hostlane2': False
 'txlos_hostlane3': False
 'txlos_hostlane4': False
 'txlos_hostlane5': False
 'txlos_hostlane6': False
 'txlos_hostlane7': False
 'txlos_hostlane8': False
 'txcdrlol_hostlane1': False
 'txcdrlol_hostlane2': False
 'txcdrlol_hostlane3': False
 'txcdrlol_hostlane4': False
 'txcdrlol_hostlane5': False
 'txcdrlol_hostlane6': False
 'txcdrlol_hostlane7': False
 'txcdrlol_hostlane8': False
 'rxlos1': False
 'rxlos2': False
 'rxlos3': False
 'rxlos4': False
 'rxlos5': False
 'rxlos6': False
 'rxlos7': False
 'rxlos8': False
 'rxcdrlol1': False
 'rxcdrlol2': False
 'rxcdrlol3': False
 'rxcdrlol4': False
 'rxcdrlol5': False
 'rxcdrlol6': False
 'rxcdrlol7': False
 'rxcdrlol8': False
 'config_state_hostlane1': 'ConfigUndefined'
 'config_state_hostlane2': 'ConfigUndefined'
 'config_state_hostlane3': 'ConfigUndefined'
 'config_state_hostlane4': 'ConfigUndefined'
 'config_state_hostlane5': 'ConfigUndefined'
 'config_state_hostlane6': 'ConfigUndefined'
 'config_state_hostlane7': 'ConfigUndefined'
 'config_state_hostlane8': 'ConfigUndefined'
 'dpinit_pending_hostlane1': False
 'dpinit_pending_hostlane2': False
 'dpinit_pending_hostlane3': False
 'dpinit_pending_hostlane4': False
 'dpinit_pending_hostlane5': False
 'dpinit_pending_hostlane6': False
 'dpinit_pending_hostlane7': False
 'dpinit_pending_hostlane8': False
 'txpowerhighalarm_flag1': False
 'txpowerlowalarm_flag1': False
 'txpowerhighwarning_flag1': False
 'txpowerlowwarning_flag1': False
 'txpowerhighalarm_flag2': False
 'txpowerlowalarm_flag2': False
 'txpowerhighwarning_flag2': False
 'txpowerlowwarning_flag2': False
 'txpowerhighalarm_flag3': False
 'txpowerlowalarm_flag3': False
 'txpowerhighwarning_flag3': False
 'txpowerlowwarning_flag3': False
 'txpowerhighalarm_flag4': False
 'txpowerlowalarm_flag4': False
 'txpowerhighwarning_flag4': False
 'txpowerlowwarning_flag4': False
 'txpowerhighalarm_flag5': False
 'txpowerlowalarm_flag5': False
 'txpowerhighwarning_flag5': False
 'txpowerlowwarning_flag5': False
 'txpowerhighalarm_flag6': False
 'txpowerlowalarm_flag6': False
 'txpowerhighwarning_flag6': False
 'txpowerlowwarning_flag6': False
 'txpowerhighalarm_flag7': False
 'txpowerlowalarm_flag7': False
 'txpowerhighwarning_flag7': False
 'txpowerlowwarning_flag7': False
 'txpowerhighalarm_flag8': False
 'txpowerlowalarm_flag8': False
 'txpowerhighwarning_flag8': False
 'txpowerlowwarning_flag8': False
 'rxpowerhighalarm_flag1': False
 'rxpowerlowalarm_flag1': False
 'rxpowerhighwarning_flag1': False
 'rxpowerlowwarning_flag1': False
 'rxpowerhighalarm_flag2': False
 'rxpowerlowalarm_flag2': False
 'rxpowerhighwarning_flag2': False
 'rxpowerlowwarning_flag2': False
 'rxpowerhighalarm_flag3': False
 'rxpowerlowalarm_flag3': False
 'rxpowerhighwarning_flag3': False
 'rxpowerlowwarning_flag3': False
 'rxpowerhighalarm_flag4': False
 'rxpowerlowalarm_flag4': False
 'rxpowerhighwarning_flag4': False
 'rxpowerlowwarning_flag4': False
 'rxpowerhighalarm_flag5': False
 'rxpowerlowalarm_flag5': False
 'rxpowerhighwarning_flag5': False
 'rxpowerlowwarning_flag5': False
 'rxpowerhighalarm_flag6': False
 'rxpowerlowalarm_flag6': False
 'rxpowerhighwarning_flag6': False
 'rxpowerlowwarning_flag6': False
 'rxpowerhighalarm_flag7': False
 'rxpowerlowalarm_flag7': False
 'rxpowerhighwarning_flag7': False
 'rxpowerlowwarning_flag7': False
 'rxpowerhighalarm_flag8': False
 'rxpowerlowalarm_flag8': False
 'rxpowerhighwarning_flag8': False
 'rxpowerlowwarning_flag8': False
 'txbiashighalarm_flag1': False
 'txbiaslowalarm_flag1': False
 'txbiashighwarning_flag1': False
 'txbiaslowwarning_flag1': False
 'txbiashighalarm_flag2': False
 'txbiaslowalarm_flag2': False
 'txbiashighwarning_flag2': False
 'txbiaslowwarning_flag2': False
 'txbiashighalarm_flag3': False
 'txbiaslowalarm_flag3': False
 'txbiashighwarning_flag3': False
 'txbiaslowwarning_flag3': False
 'txbiashighalarm_flag4': False
 'txbiaslowalarm_flag4': False
 'txbiashighwarning_flag4': False
 'txbiaslowwarning_flag4': False
 'txbiashighalarm_flag5': False
 'txbiaslowalarm_flag5': False
 'txbiashighwarning_flag5': False
 'txbiaslowwarning_flag5': False
 'txbiashighalarm_flag6': False
 'txbiaslowalarm_flag6': False
 'txbiashighwarning_flag6': False
 'txbiaslowwarning_flag6': False
 'txbiashighalarm_flag7': False
 'txbiaslowalarm_flag7': False
 'txbiashighwarning_flag7': False
 'txbiaslowwarning_flag7': False
 'txbiashighalarm_flag8': False
 'txbiaslowalarm_flag8': False
 'txbiashighwarning_flag8': False
 'txbiaslowwarning_flag8': False
 'prefecberhighalarm_flag': False
 'prefecberlowalarm_flag': True
 'prefecberhighwarning_flag': False
 'prefecberlowwarning_flag': False}



get_transceiver_loopback

{'simultaneous_host_media_loopback_supported': True
 'per_lane_media_loopback_supported': True
 'per_lane_host_loopback_supported': True
 'host_side_input_loopback_supported': True
 'host_side_output_loopback_supported': True
 'media_side_input_loopback_supported': True
 'media_side_output_loopback_supported': True
 'media_output_loopback': False
 'media_input_loopback': False
 'host_output_loopback_lane1': False
 'host_output_loopback_lane2': False
 'host_output_loopback_lane3': False
 'host_output_loopback_lane4': False
 'host_output_loopback_lane5': False
 'host_output_loopback_lane6': False
 'host_output_loopback_lane7': False
 'host_output_loopback_lane8': False
 'host_input_loopback_lane1': False
 'host_input_loopback_lane2': False
 'host_input_loopback_lane3': False
 'host_input_loopback_lane4': False
 'host_input_loopback_lane5': False
 'host_input_loopback_lane6': False
 'host_input_loopback_lane7': False
 'host_input_loopback_lane8': False}

aravindmani-1 commented on sonic_platform_base/sonic_xcvr/codes/public/sff8024.py:
```diff
@@ -345,26 +345,26 @@ class Sff8024(XcvrCodes):
+        0: 'Undefined',
+        1: 'Active Cable assembly with BER < 10^-12',
+        2: 'Active Cable assembly with BER < 5x10^-5',
         '03': 'Active Cable assembly with BER < 2.6x10^-4',
```
Don't we have to change here also?.

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -180,24 +180,24 @@ def get_transceiver_bulk_status(self):
-            bulk_status["tx%dbias" % i] = tx_bias['LaserBiasTx%dField' % i]
-            bulk_status["rx%dpower" % i] = rx_power['OpticalPowerRx%dField' %i]
-            bulk_status["tx%dpower" % i] = tx_power['OpticalPowerTx%dField' %i]
+            bulk_status["tx%dbias" % i] = tx_bias['LaserBiasTx%dField' % i] if self.get_tx_bias_support() else 'N/A'
```
There are two things we need to check:
1. Module is flat memory module
2. Module is non-flat memory module but still doesn't support tx bias, rx power, bias current monitoring (See page 01, byte 160)

Please modify the get_tx_bias_support as:
```
def get_tx_bias_support():
   if is not self.flat_memory() and page1, byte 160, bit 0 is set:
        return True
    return false
```

qinchuanares replied:
done

prgeor replied:
get_tx_bias_support() is already present inside get_tx_bias() so this extra if check is not required here.

qinchuanares replied:
it will return a list of 'N/A' if not supported. But tx_bias['LaserBiasTx%dField' % i] will return KeyError. That's why I add this if else statement in this line.

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -180,24 +180,24 @@ def get_transceiver_bulk_status(self):
-            bulk_status["rx%dpower" % i] = rx_power['OpticalPowerRx%dField' %i]
-            bulk_status["tx%dpower" % i] = tx_power['OpticalPowerTx%dField' %i]
+            bulk_status["tx%dbias" % i] = tx_bias['LaserBiasTx%dField' % i] if self.get_tx_bias_support() else 'N/A'
+            bulk_status["rx%dpower" % i] = rx_power['OpticalPowerRx%dField' %i] if self.get_tx_power_support() else 'N/A'
+            bulk_status["tx%dpower" % i] = tx_power['OpticalPowerTx%dField' %i] if self.get_rx_power_support() else 'N/A'
```
please modify the definition of get_tx_power_support() and get_rx_power_support() as mentioned in above comment for get_tx_bias_support() API. Also, rx and tx APIs are swapped above.

prgeor replied:
if you see the definition of  get_tx_power(), it should return N/A if its not supported, so there is no need to put checks in line 191, 192. so fix the api get_tx_power() and it should automatically fix here.

qinchuanares replied:
modified the support function. For line 191 and 192, I am calling keys in the dictionary but they are non-existent if it returns N/A

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -895,11 +898,16 @@ def set_loopback_mode(self, loopback_mode):
         self.cdb = CmisCdbApi(self.xcvr_eeprom)
         return self.cdb
 
+    def get_vdm_support(self):
```
Add additional check page1, byte 142, bit 6

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1642,6 +1653,9 @@ def get_transceiver_loopback(self):
+        if loopback_support is None:
+            return None
         trans_loopback = dict()
         trans_loopback['media_output_loopback'] = self.get_media_output_loopback()
```
please add check for get_media_output_loopback() and other loopback APIs by checking the advertisement in page13, byte128

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1642,6 +1653,9 @@ def get_transceiver_loopback(self):
         host_input_loopback_lane8   = BOOLEAN                          ; host side input loopback enable lane8
         ========================================================================
         """
+        loopback_support = self.get_transceiver_loopback_support()
```
This check is not required if you add the check inside the different loopback APIs, see comments below

qinchuanares replied:
Need to check if it is flat memory, otherwise won't be able to read page 0x13

prgeor replied:
i meant we can have one single API get_loopback_capability() which will check if its flat memory and also check other loopback capability. we don't need get_transceiver_loopback_support() and get_loopback_capability()

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1642,6 +1653,9 @@ def get_transceiver_loopback(self):
+            return None
         trans_loopback = dict()
         trans_loopback['media_output_loopback'] = self.get_media_output_loopback()
         trans_loopback['media_input_loopback'] = self.get_media_input_loopback()
```
if the loopack is not supported, return a list of 'N/A' for each lane.

qinchuanares replied:
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/codes/public/sff8024.py:
```diff
@@ -345,26 +345,26 @@ class Sff8024(XcvrCodes):
+        0: 'Undefined',
+        1: 'Active Cable assembly with BER < 10^-12',
+        2: 'Active Cable assembly with BER < 5x10^-5',
         '03': 'Active Cable assembly with BER < 2.6x10^-4',
         '04': 'Active Cable assembly with BER < 10^-6',
         'bf': 'Active Loopback module'
```
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -241,10 +242,10 @@ def get_transceiver_threshold_info(self):
-            "txbiaslowalarm": float("{:.3f}".format(thresh[consts.TX_BIAS_LOW_ALARM_FIELD])),
-            "txbiashighwarning": float("{:.3f}".format(thresh[consts.TX_BIAS_HIGH_WARNING_FIELD])),
-            "txbiaslowwarning": float("{:.3f}".format(thresh[consts.TX_BIAS_LOW_WARNING_FIELD]))
+            "txbiashighalarm": float("{:.3f}".format(thresh[consts.TX_BIAS_HIGH_ALARM_FIELD]*tx_bias_scale)),
+            "txbiaslowalarm": float("{:.3f}".format(thresh[consts.TX_BIAS_LOW_ALARM_FIELD]*tx_bias_scale)),
+            "txbiashighwarning": float("{:.3f}".format(thresh[consts.TX_BIAS_HIGH_WARNING_FIELD]*tx_bias_scale)),
+            "txbiaslowwarning": float("{:.3f}".format(thresh[consts.TX_BIAS_LOW_WARNING_FIELD]*tx_bias_scale))
```
please use dependency field, so that the decoding logic is all at one place (not inside APIs).

qinchuanares replied:
the scale is in Page1 Byte 160 Bit3-4 and having this conversion:

0b: multiply x1
01b: multiply x2
10b: multiply x4
11b: reserved 

It is not easy to include it in the field as a scale argument.

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -476,7 +481,7 @@ def get_tx_power(self):
 
     def get_tx_power_support(self):
-        return not self.is_flat_memory()
+        return not self.is_flat_memory() and self.xcvr_eeprom.read(consts.TX_POWER_SUPPORT_FIELD)
```
i don't think you can read this field which is inside a reg-group field directly

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -491,7 +496,7 @@ def get_rx_power(self):
 
     def get_rx_power_support(self):
-        return not self.is_flat_memory()
+        return not self.is_flat_memory() and self.xcvr_eeprom.read(consts.RX_POWER_SUPPORT_FIELD)
```
this field is inside reg group MODULE_CHAR_ADVT_FIELD so cannot be read directly.

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -513,7 +518,7 @@ def get_tx_fault(self):
 
     def get_tx_los_support(self):
-        return not self.is_flat_memory()
+        return not self.is_flat_memory() and self.xcvr_eeprom.read(consts.TX_LOS_SUPPORT_FIELD)
```
this field is inside reg group MODULE_CHAR_ADVT_FIELD so cannot be read directly.

prgeor commented:
@qinchuanares the expected return type for get_rx_los() should return list because [sonic-mgmt test](https://github.com/Azure/sonic-mgmt/blob/master/tests/platform_tests/api/test_sfp.py#L323) expects it to be of list type. Please fix other APIs as well.

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -181,24 +181,24 @@ def get_transceiver_bulk_status(self):
         bulk_status = {
             "rx_los": all(rx_los.values()) if self.get_rx_los_support() else 'N/A',
             "tx_fault": all(tx_fault.values()) if self.get_tx_fault_support() else 'N/A',
-            "tx_disable": all(tx_disable),
+            "tx_disable": all(tx_disable) if self.get_tx_disable_support() else 'N/A',
             "tx_disabled_channel": tx_disabled_channel,
```
These are per lane, please add per lane status

qinchuanares replied:
done

qinchuanares commented:
> @qinchuanares the expected return type for get_rx_los() should return list because [sonic-mgmt test](https://github.com/Azure/sonic-mgmt/blob/master/tests/platform_tests/api/test_sfp.py#L323) expects it to be of list type. Please fix other APIs as well.

changed output from dictionary to list
function list:
1. get_tx_cdr_lol()
2. get_rx_los()
3. get_rx_cdr_lol()
4. get_tx_fault()
5. get_tx_los()

I also changed get_transceiver_info() and get_transceiver_status() accordingly to adpat to the changes in the aforementioned functions.

prgeor commented:
@qinchuanares could you resolve few conflicts?

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1624,33 +1595,96 @@ def get_transceiver_status(self):
+            trans_status['txlos_hostlane7'] = tx_los[6]
+            trans_status['txlos_hostlane8'] = tx_los[7]
+            tx_lol = self.get_tx_cdr_lol()
+            trans_status['txcdrlol_hostlane1'] = tx_lol[0]
+            trans_status['txcdrlol_hostlane2'] = tx_lol[1]
+            trans_status['txcdrlol_hostlane3'] = tx_lol[2]
+            trans_status['txcdrlol_hostlane4'] = tx_lol[3]
+            trans_status['txcdrlol_hostlane5'] = tx_lol[4]
+            trans_status['txcdrlol_hostlane6'] = tx_lol[5]
+            trans_status['txcdrlol_hostlane7'] = tx_lol[6]
+            trans_status['txcdrlol_hostlane8'] = tx_lol[7]
```
can we condense these line to few lines?

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1624,33 +1595,96 @@ def get_transceiver_status(self):
+            tx_fault = self.get_tx_fault()
+            trans_status['txfault'] = tx_fault[0]
+            tx_los = self.get_tx_los()
+            trans_status['txlos_hostlane1'] = tx_los[0]
+            trans_status['txlos_hostlane2'] = tx_los[1]
+            trans_status['txlos_hostlane3'] = tx_los[2]
+            trans_status['txlos_hostlane4'] = tx_los[3]
+            trans_status['txlos_hostlane5'] = tx_los[4]
+            trans_status['txlos_hostlane6'] = tx_los[5]
+            trans_status['txlos_hostlane7'] = tx_los[6]
+            trans_status['txlos_hostlane8'] = tx_los[7]
```
can we condense these lines to few lines?

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1624,33 +1595,96 @@ def get_transceiver_status(self):
+            rx_lol = self.get_rx_cdr_lol()
+            trans_status['rxcdrlol'] = rx_lol[0]
+            config_status_dict = self.get_config_datapath_hostlane_status()
+            trans_status['config_state_hostlane1'] = config_status_dict['ConfigStatusLane1']
+            trans_status['config_state_hostlane2'] = config_status_dict['ConfigStatusLane2']
+            trans_status['config_state_hostlane3'] = config_status_dict['ConfigStatusLane3']
+            trans_status['config_state_hostlane4'] = config_status_dict['ConfigStatusLane4']
+            trans_status['config_state_hostlane5'] = config_status_dict['ConfigStatusLane5']
+            trans_status['config_state_hostlane6'] = config_status_dict['ConfigStatusLane6']
+            trans_status['config_state_hostlane7'] = config_status_dict['ConfigStatusLane7']
+            trans_status['config_state_hostlane8'] = config_status_dict['ConfigStatusLane8']
```
Please condense these lines to few lines

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1624,33 +1595,96 @@ def get_transceiver_status(self):
+            trans_status['config_state_hostlane7'] = config_status_dict['ConfigStatusLane7']
+            trans_status['config_state_hostlane8'] = config_status_dict['ConfigStatusLane8']
+            dpinit_pending_dict = self.get_dpinit_pending()
+            trans_status['dpinit_pending_hostlane1'] = dpinit_pending_dict['DPInitPending1']
+            trans_status['dpinit_pending_hostlane2'] = dpinit_pending_dict['DPInitPending2']
+            trans_status['dpinit_pending_hostlane3'] = dpinit_pending_dict['DPInitPending3']
+            trans_status['dpinit_pending_hostlane4'] = dpinit_pending_dict['DPInitPending4']
+            trans_status['dpinit_pending_hostlane5'] = dpinit_pending_dict['DPInitPending5']
+            trans_status['dpinit_pending_hostlane6'] = dpinit_pending_dict['DPInitPending6']
+            trans_status['dpinit_pending_hostlane7'] = dpinit_pending_dict['DPInitPending7']
+            trans_status['dpinit_pending_hostlane8'] = dpinit_pending_dict['DPInitPending8']
```
can we condense these lines to few lines?

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1624,33 +1595,96 @@ def get_transceiver_status(self):
-            pass
+        if not self.is_flat_memory():
+            dp_state_dict = self.get_datapath_state()
+            trans_status['DP1State'] = dp_state_dict['DP1State']
+            trans_status['DP2State'] = dp_state_dict['DP2State']
+            trans_status['DP3State'] = dp_state_dict['DP3State']
+            trans_status['DP4State'] = dp_state_dict['DP4State']
+            trans_status['DP5State'] = dp_state_dict['DP5State']
+            trans_status['DP6State'] = dp_state_dict['DP6State']
+            trans_status['DP7State'] = dp_state_dict['DP7State']
```
please condense these to few lines

qinchuanares replied:
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1644,25 +1681,87 @@ def get_transceiver_loopback(self):
+            trans_loopback['media_side_output_loopback_supported'] = 'N/A'
+            trans_loopback['media_output_loopback'] = 'N/A'
+            trans_loopback['media_input_loopback'] = 'N/A'
+            trans_loopback['host_output_loopback_lane1'] = 'N/A'
+            trans_loopback['host_output_loopback_lane2'] = 'N/A'
+            trans_loopback['host_output_loopback_lane3'] = 'N/A'
+            trans_loopback['host_output_loopback_lane4'] = 'N/A'
+            trans_loopback['host_output_loopback_lane5'] = 'N/A'
+            trans_loopback['host_output_loopback_lane6'] = 'N/A'
+            trans_loopback['host_output_loopback_lane7'] = 'N/A'
+            trans_loopback['host_output_loopback_lane8'] = 'N/A'
```
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1644,25 +1681,87 @@ def get_transceiver_loopback(self):
+            trans_loopback['host_output_loopback_lane6'] = 'N/A'
+            trans_loopback['host_output_loopback_lane7'] = 'N/A'
+            trans_loopback['host_output_loopback_lane8'] = 'N/A'
+            trans_loopback['host_input_loopback_lane1'] = 'N/A'
+            trans_loopback['host_input_loopback_lane2'] = 'N/A'
+            trans_loopback['host_input_loopback_lane3'] = 'N/A'
+            trans_loopback['host_input_loopback_lane4'] = 'N/A'
+            trans_loopback['host_input_loopback_lane5'] = 'N/A'
+            trans_loopback['host_input_loopback_lane6'] = 'N/A'
+            trans_loopback['host_input_loopback_lane7'] = 'N/A'
+            trans_loopback['host_input_loopback_lane8'] = 'N/A'
```
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1644,25 +1681,87 @@ def get_transceiver_loopback(self):
+            trans_loopback['host_input_loopback_lane7'] = host_input_loopback_status[6]
+            trans_loopback['host_input_loopback_lane8'] = host_input_loopback_status[7]
+        else:
+            trans_loopback['host_input_loopback_lane1'] = 'N/A'
+            trans_loopback['host_input_loopback_lane2'] = 'N/A'
+            trans_loopback['host_input_loopback_lane3'] = 'N/A'
+            trans_loopback['host_input_loopback_lane4'] = 'N/A'
+            trans_loopback['host_input_loopback_lane5'] = 'N/A'
+            trans_loopback['host_input_loopback_lane6'] = 'N/A'
+            trans_loopback['host_input_loopback_lane7'] = 'N/A'
+            trans_loopback['host_input_loopback_lane8'] = 'N/A'
```
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1644,25 +1681,87 @@ def get_transceiver_loopback(self):
+            trans_loopback['host_output_loopback_lane8'] = 'N/A'
+        if loopback_capability['host_side_input_loopback_supported']:
+            host_input_loopback_status = self.get_host_input_loopback()
+            trans_loopback['host_input_loopback_lane1'] = host_input_loopback_status[0]
+            trans_loopback['host_input_loopback_lane2'] = host_input_loopback_status[1]
+            trans_loopback['host_input_loopback_lane3'] = host_input_loopback_status[2]
+            trans_loopback['host_input_loopback_lane4'] = host_input_loopback_status[3]
+            trans_loopback['host_input_loopback_lane5'] = host_input_loopback_status[4]
+            trans_loopback['host_input_loopback_lane6'] = host_input_loopback_status[5]
+            trans_loopback['host_input_loopback_lane7'] = host_input_loopback_status[6]
+            trans_loopback['host_input_loopback_lane8'] = host_input_loopback_status[7]
```
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1644,25 +1681,87 @@ def get_transceiver_loopback(self):
+            trans_loopback['host_output_loopback_lane7'] = host_output_loopback_status[6]
+            trans_loopback['host_output_loopback_lane8'] = host_output_loopback_status[7]
+        else:
+            trans_loopback['host_output_loopback_lane1'] = 'N/A'
+            trans_loopback['host_output_loopback_lane2'] = 'N/A'
+            trans_loopback['host_output_loopback_lane3'] = 'N/A'
+            trans_loopback['host_output_loopback_lane4'] = 'N/A'
+            trans_loopback['host_output_loopback_lane5'] = 'N/A'
+            trans_loopback['host_output_loopback_lane6'] = 'N/A'
+            trans_loopback['host_output_loopback_lane7'] = 'N/A'
+            trans_loopback['host_output_loopback_lane8'] = 'N/A'
```
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1644,25 +1681,87 @@ def get_transceiver_loopback(self):
+            trans_loopback['media_input_loopback'] = 'N/A'
+        if loopback_capability['host_side_output_loopback_supported']:
+            host_output_loopback_status = self.get_host_output_loopback()
+            trans_loopback['host_output_loopback_lane1'] = host_output_loopback_status[0]
+            trans_loopback['host_output_loopback_lane2'] = host_output_loopback_status[1]
+            trans_loopback['host_output_loopback_lane3'] = host_output_loopback_status[2]
+            trans_loopback['host_output_loopback_lane4'] = host_output_loopback_status[3]
+            trans_loopback['host_output_loopback_lane5'] = host_output_loopback_status[4]
+            trans_loopback['host_output_loopback_lane6'] = host_output_loopback_status[5]
+            trans_loopback['host_output_loopback_lane7'] = host_output_loopback_status[6]
+            trans_loopback['host_output_loopback_lane8'] = host_output_loopback_status[7]
```
done

qinchuanares commented on sonic_platform_base/sonic_xcvr/sfp_optoe_base.py:
```diff
@@ -31,6 +31,50 @@ def get_transceiver_threshold_info(self):
+        api = self.get_xcvr_api()
+        return api.get_module_fw_upgrade_feature() if api is not None else None
+
+    def get_module_fw_info(self):
+        api = self.get_xcvr_api()
+        return api.get_module_fw_info() if api is not None else None
+
+    def module_fw_run(self):
+        api = self.get_xcvr_api()
+        return api.module_fw_run() if api is not None else None
+
+    def module_fw_commit(self):
+        api = self.get_xcvr_api()
+        return api.module_fw_commit() if api is not None else None
+
+    def module_fw_download(self):
+        api = self.get_xcvr_api()
+        return api.module_fw_download() if api is not None else None
+
+    def module_fw_upgrade(self):
+        api = self.get_xcvr_api()
+        return api.module_fw_upgrade() if api is not None else None
+
+    def module_fw_switch(self):
+        api = self.get_xcvr_api()
+        return api.module_fw_switch() if api is not None else None
```
can we leave these functions for testing purpose?

prgeor replied:
for testing purspose we can directly call the api. No need for making changes inside sfp_optoe_base class which is used by SFP/QSFP/QSFP-DD/OSFP

qinchuanares commented on sonic_platform_base/sonic_xcvr/sfp_optoe_base.py:
```diff
@@ -31,6 +31,50 @@ def get_transceiver_threshold_info(self):
+        api = self.get_xcvr_api()
+        return api.get_transceiver_pm() if api is not None else None
+    
+    def get_module_fw_upgrade_feature(self):
+        api = self.get_xcvr_api()
+        return api.get_module_fw_upgrade_feature() if api is not None else None
```
can we leave these functions for testing purpose?

prgeor replied:
for testing purspose we can directly call the api. No need for making changes inside sfp_optoe_base class which is used by SFP/QSFP/QSFP-DD/OSFP

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1647,33 +1611,60 @@ def get_transceiver_status(self):
+            for lane in range(1, self.NUM_CHANNELS+1):
+                trans_status['txcdrlol_hostlane%d' % lane] = tx_lol[lane - 1]
+            rx_los = self.get_rx_los()
+            trans_status['rxlos'] = rx_los[0]
```
cmis.py is applicable for non-ZR optics, so why only first lane?

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1647,33 +1611,60 @@ def get_transceiver_status(self):
+            rx_los = self.get_rx_los()
+            trans_status['rxlos'] = rx_los[0]
+            rx_lol = self.get_rx_cdr_lol()
+            trans_status['rxcdrlol'] = rx_lol[0]
```
what about lane 1 to 8 for non-ZR?

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1647,33 +1611,60 @@ def get_transceiver_status(self):
+            trans_status['txpowerhighwarning_flag'] = tx_power_flag_dict['tx_power_high_warn']['TxPowerHighWarnFlag1']
+            trans_status['txpowerlowwarning_flag'] = tx_power_flag_dict['tx_power_low_warn']['TxPowerLowWarnFlag1']
+            rx_power_flag_dict = self.get_rx_power_flag()
+            trans_status['rxpowerhighalarm_flag'] = rx_power_flag_dict['rx_power_high_alarm']['RxPowerHighAlarmFlag1']
+            trans_status['rxpowerlowalarm_flag'] = rx_power_flag_dict['rx_power_low_alarm']['RxPowerLowAlarmFlag1']
+            trans_status['rxpowerhighwarning_flag'] = rx_power_flag_dict['rx_power_high_warn']['RxPowerHighWarnFlag1']
+            trans_status['rxpowerlowwarning_flag'] = rx_power_flag_dict['rx_power_low_warn']['RxPowerLowWarnFlag1']
```
what about other lanes for non-ZR?

qinchuanares replied:
done

prgeor commented on sonic_platform_base/sonic_xcvr/api/public/cmis.py:
```diff
@@ -1647,33 +1611,60 @@ def get_transceiver_status(self):
+            trans_status['rxpowerhighwarning_flag'] = rx_power_flag_dict['rx_power_high_warn']['RxPowerHighWarnFlag1']
+            trans_status['rxpowerlowwarning_flag'] = rx_power_flag_dict['rx_power_low_warn']['RxPowerLowWarnFlag1']
+            tx_bias_flag_dict = self.get_tx_bias_flag()
+            trans_status['txbiashighalarm_flag'] = tx_bias_flag_dict['tx_bias_high_alarm']['TxBiasHighAlarmFlag1']
+            trans_status['txbiaslowalarm_flag'] = tx_bias_flag_dict['tx_bias_low_alarm']['TxBiasLowAlarmFlag1']
+            trans_status['txbiashighwarning_flag'] = tx_bias_flag_dict['tx_bias_high_warn']['TxBiasHighWarnFlag1']
+            trans_status['txbiaslowwarning_flag'] = tx_bias_flag_dict['tx_bias_low_warn']['TxBiasLowWarnFlag1']
```
what about other lanes for non-ZR?

qinchuanares replied:
done
